calliope/backend/parsing.py

```diff
@@ -0,0 +1,104 @@
+import itertools
+from typing import Dict, KeysView, List
+
+from typing_extensions import NotRequired, TypedDict, Required
+
+import pyparsing as pp
+import xarray as xr
+
+
+class ForEach(TypedDict):
+    set_item: str
+    set_name: str
+
+
+# TODO: decide if this should simply be typed as a Dict and all the details are left in a YAML schema
```

Comment:
Actually, maybe this is merged with `ParsedConstraint` so that you have one object that essentially parses itself, adding additional parsed properties after it has done so, but keeping the original "raw" data around too?

Reply:
What are you referring to here? My TODO is about the type hinting (i.e.: https://peps.python.org/pep-0589/)